[oss][pytorch] Add quint2x4 dtype

ShijunK opened the pull request:
Summary:
Introduce 2bit qtensor. The new dtype added for this is c10::quint2x4

The underlying storage for this is still uint8_t, so we pack 4 2-bit values in a byte while quantizing it.

Kernels that use this dtype should be aware of the packing format. (4 2-bit values in one byte)

Test Plan: `buck test mode/dev-asan caffe2/test/:quantization -- test_qtensor`

Differential Revision: D31148141

ezyang commented on aten/src/TH/THGenerateQUInt2x4Type.h:
```diff
@@ -0,0 +1,24 @@
+
+#ifndef THGenerateManyTypes
+#undef TH_GENERIC_FILE
+#endif
```
@ngimel this is my unhappy face 🧐

supriyar commented on aten/src/ATen/quantized/Quantizer.cpp:
```diff
@@ -86,6 +86,9 @@ int64_t get_sub_byte_tensor_size(int64_t size_bytes, at::ScalarType t) {
     case at::ScalarType::QUInt4x2:
       new_size_bytes = std::ceil(size_bytes * 0.5);
       break;
+    case at::ScalarType::QUInt2x4:
+      new_size_bytes = std::ceil(size_bytes * 0.25);
```
Not related to this PR but I think this code has a bug. 
for 4x2 type num bytes its celing function But it should be applied for the innermost dimension.
So if you have tensor of size say 31x31 for 4 bit you would want to allocate 16 bytes per row Thus total of 16 x 31 = 496 bytes. But if you do 31 x 31 = 961 bytes then ceil(961 * 0.5) = 481 bytes.

Since we are modifying this and adding support 2bits, can we also fix this?

ezyang replied:
While you're at it, can you use one of the idioms for round to ceil integer division that doesn't go through floating point. If your numbers here are big the floating point inaccuracy will also be a problem.

ngimel replied:
at::ceil_div(size_bytes, 4)

ShijunK replied:
@supriyar , @ezyang could you have another look after latest update?

ezyang reviewed (changes requested):
TESTS???

ShijunK commented on aten/src/ATen/native/quantized/cpu/int_repr_quant.cpp:
```diff
@@ -14,8 +14,8 @@ Tensor int_repr_quantized_cpu(const Tensor& self) {
-    if (bit_width == 4) {
-      int64_t out_size = std::ceil(self.numel() * 0.5);
+    if (bit_width == 4 || bit_width == 2) {
+      int64_t out_size = std::ceil(self.numel() * bit_width / 8);
```
good catch, thx @ezyang  I make mistake assuming that existing tests should have cover the scenario. 

added new tests "test_qtensor_int_repr", "test_qtensor_sub_byte_aligned_cols" and "test_qtensor_sub_byte_not_aligned_cols" under "test/quantization/core/test_quantized_tensor.py"

ShijunK commented:
> TESTS???

added new tests "test_qtensor_int_repr", "test_qtensor_sub_byte_aligned_cols" and "test_qtensor_sub_byte_not_aligned_cols" under "test/quantization/core/test_quantized_tensor.py"

cc @ezyang @supriyar

supriyar reviewed (approved):
lg, thanks!

ShijunK commented:
rebase after https://github.com/pytorch/pytorch/pull/62030

ShijunK committed:
[oss][pytorch] Add quint2x4 dtype (#65545)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/65545

Introduce 2bit qtensor. The new dtype added for this is c10::quint2x4

The underlying storage for this is still uint8_t, so we pack 4 2-bit values in a byte while quantizing it.

Kernels that use this dtype should be aware of the packing format. (4 2-bit values in one byte)

Test Plan: `buck test mode/dev-asan caffe2/test/:quantization -- test_qtensor`

Reviewed By: supriyar

Differential Revision: D31148141

fbshipit-source-id: abda24f0d81445acf89d8d4cde667bc65cdd3d22
